feat(flexslider): no opacity anymore

Mte90 opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Remove the opacity to the gallery that is reported as CLS from Lighthouse and is an old way to hide the gallery lazyload. Now WP and browser have a native way for images and a lot of things are happened in the last 6 years. Also after this https://github.com/woocommerce/woocommerce/pull/31407 some information are now in the page to improve the layout and size of the gallery.
The only doubts is that some CSS was inline in the templates folder, so as is removed now for all the themes around that copied still have this CSS. With those changes the opacity is not removed, so it is required a workaround to flip that value as true if opacity is set. Before to do that I am looking for feedbacks.

Closes #30893.

This changes is by me and @erikyo as we talked about how to do it in the best way.

### How to test the changes in this Pull Request:

1. Go to single product with Chrome
2. Check the CLS

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Mte90 committed:
feat(flexslider): no opacity anymore

Mte90 commented on plugins/woocommerce/legacy/js/frontend/single-product.js:
```diff
@@ -93,7 +93,6 @@ jQuery( function( $ ) {
 
 		// No images? Abort.
 		if ( 0 === this.$images.length ) {
-			this.$target.css( 'opacity', 1 );
```
Maybe we can move this line outside the if to let oldest template to work with this changes.

Mte90 committed:
check if there is opacity and show

jeffstieler reviewed (changes requested):
Thank you for your contribution @Mte90!

This looks good on products with only one image, but the opacity changes cause other layout shifts on products with multiple images (while images are loading). See this screen capture of a page refresh:

![Capture 2022-01-18 at 09 57 43](https://user-images.githubusercontent.com/63922/149983355-cb529d77-d9fa-4d49-b03e-5dc09c159ee7.gif)

Mte90 commented:
So probably hide all the images except the first one should fix that user case. What do you think @erikyo?

erikyo commented:
@jeffstieler maybe the [patch](https://github.com/woocommerce/woocommerce/blob/c9696e66b9ebe1adaf6bf8a0f28fe80f38dae220/plugins/woocommerce/legacy/js/flexslider/jquery.flexslider.js#L247) that set the with and the height of the thumbnail images isn't present in your test environment? In my blog this fix work as it should, check [this](https://modul-r.codekraft.it/product/logo-collection/)

jeffstieler commented:
@erikyo looks like you've identified the root cause of the glitch. Thanks again for taking the time to make these fixes. (My local test environment did have the latest patch to `jquery.flexslider.js`, fwiw)

Mte90 committed:
move doMath to avoid misconfigured size during loading

erikyo commented:
@jeffstieler I'm sorry but what I said above is inaccurate and after further investigation seems the issue was a little more complex and related in how the slideshow and its nav are created... I try explain better with some screenshots the loading sequence:
![image](https://user-images.githubusercontent.com/8550908/150142200-bf34316a-fc16-4fdd-9ddc-13702d332170.png)
1) before the flexslider is loaded ($( 'body' ).on( 'init'..) the scenario is this, the first image is displayed at 100% of the container width and the following slides at 25% of the width. There is not yet the "real" nav, in fact the first image is missing.
![image](https://user-images.githubusercontent.com/8550908/150143116-4577d609-74fc-40bb-9229-38c7d18fea05.png)
2) then the slider images are resized to 100% of the width
![image](https://user-images.githubusercontent.com/8550908/150144517-58009c55-eb86-45c6-87e6-ad6d71ab9692.png)
3) the images after being resized are aligned to the first one (actually this happens because Flexslider sets the width of the container)
![image](https://user-images.githubusercontent.com/8550908/150144980-ef9c424f-d685-4db9-a984-89563c0cbf89.png)
4) finally the nav is created using the data taken from the images (that's why each div has the data-thumb attribute)
---
We are looking (with @Mte90) for a solution to make these things happen without upsetting flexslider, it would be enough for the images to be moved at the moment that the nav was created and only after this apply the resize to the hidden images (slider2, slider3, slider4,...).

erikyo committed:
optimizes slideshow loading sequence / performance

erikyo committed:
add srcset to the nav thumb images so the images already loaded can be reused

Mte90 commented:
This last changes should improve the situation, @erikyo says that now doesn't load anymore 3 images at full size so LightHouse is very happy of that (and the browsers too).
Looking at the code is missing to do some changes to the fade part.

erikyo committed:
A faster loading sequence for the vertical slider.
- Set a fixed size for slider images and maybe additionally solves an images size issue for all the images next to the first one (vertical slider only)

erikyo committed:
fix slider fade transition type when nav is enabled (the height was set to the parent container)

erikyo committed:
there is no need to recalculate the maximum width any further

erikyo committed:
this creates a bit of a layout shift but in the end it's not the case to overcomplicate things

erikyo committed:
fix to correctly display the thumbnails per row set with the filter "woocommerce_product_thumbnails_columns".
After this commit the heights and widths of the nav thumbs are pre-calculated and forced to square shape.

erikyo committed:
Merge branch 'trunk' into flexslider-no-opacity

erikyo commented:
This makes sure that the thumbnails that are set with the `woocommerce_product_thumbnails_columns` filter are displayed correctly. Moreover this commit will speed up loading because the width and the height of thumbnails are pre-calculated using the number of columns (which [was there](https://github.com/woocommerce/woocommerce/blob/b5ae232106cd5d0d13869a0b030f0f30e5cd3e53/plugins/woocommerce/templates/single-product/product-image.php#L39) but not used). 
Thanks to srcset under certain conditions are "reclycled" from the larger image and the browser may opt not to download the thumbnail using the sharper image. This has a huge benefit with lighthouse/mobile rating.

`add_filter( 'woocommerce_product_thumbnails_columns', function () { return 5; } ); // 3,4,5 are allowed values`

Those using non-square thumbnails might see the image not filling its box, but this could be fixed easily setting a fixed height to navigation items, with something like:
```
body.woocommerce div.product div.images .flex-control-thumbs li {
    height: 100px;
    margin: 10px auto;
}
.woocommerce div.product div.images .flex-control-thumbs li img {
    height: 100%;
    object-fit: contain;
    /* object-position: left; */
}
```

erikyo committed:
prevents doMath from return wrong calcs during resizing

Mte90 commented:
There is any opportunity to get this change deployed? 
My customers are asking to improve the performances and I can't do anything as it is a Woocommerce fault and they don't understand that is not something we can do until this is merged (yes I can dequeue the script etc) but I can't do it for all my customers.

peterfabian commented:
A follow-up PR is being created by @erikyo atm, so please don't review this one just yet. Thx!

Mte90 commented:
So I think that this can be closed for https://github.com/woocommerce/woocommerce/pull/33041
